aiohttp/client_reqrep.py

```diff
@@ -1022,7 +1023,12 @@ def raise_for_status(self) -> None:
         if not self.ok:
             # reason should always be not None for a started response
             assert self.reason is not None
-            self.release()
+
+            # If we're in a context we can rely on __aexit__() to release as the
```

Comment:
Does this mean we *want to* support/tolerate cases where the callers don't use CMs? I'd say this encourages an antipattern. I'd rather delete the auto-release here in favor or telling people to always use a CM or release themselves…

Reply:
This has been like this for some time. My reasoning for why this is the expected behaviour is that if they have code to release it themselves, it may not get run because we're throwing an exception here that could cause it to bypass that code. So, instead we make this an explicit release for this exceptional case.

Reply:
But if we expected that people were to do the right thing and always use the async context manager, that wouldn't be a problem, right?

Reply:
We do explicitly allow `await` instead of the context manager in our code and documentation, even if we generally discourage it. We also have a substantial number of tests validating such behaviour (e.g. validating that .close() is not needed after a .read() call etc.).

If we wanted to remove the `await` usage, that seems like an entirely different discussion and could only go into v4.

Reply:
Fair enough. Marking the thread as resolved.